lib/cloud/aws.go

```diff
@@ -0,0 +1,168 @@
+	// instance of awsConfigCacheKey.
+	configsCache *utils.FnCache
+}
+
+func newAWSClientsV2() (*awsClientsV2, error) {
+	configsCache, err := utils.NewFnCache(utils.FnCacheConfig{
+		TTL: 15 * time.Minute,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &awsClientsV2{
+		configsCache: configsCache,
+	}, nil
+}
+
+// GetAWSConfigV2 returns AWS config for the specified region, optionally
+// assuming AWS IAM Roles.
+func (c *awsClientsV2) GetAWSConfigV2(ctx context.Context, region string, opts ...AWSOptionsFn) (*aws.Config, error) {
+	var options awsOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+	var err error
+	if options.baseConfigV2 == nil {
+		options.baseConfigV2, err = c.getAWSConfigV2ForRegion(ctx, region, options)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	if options.assumeRoleARN == "" {
+		return options.baseConfigV2, nil
+	}
+	return c.getAWSConfigV2ForRole(ctx, region, options)
+}
+
+// GetAWSSTSClientV2 returns AWS STS client for the specified region.
+func (c *awsClientsV2) GetAWSSTSClientV2(ctx context.Context, region string, opts ...AWSOptionsFn) (cloudaws.STSAPI, error) {
+	config, err := c.GetAWSConfigV2(ctx, region, opts...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return cloudaws.NewSTSClient(*config), nil
+}
+
+// getAWSConfigV2ForRegion returns AWS config for the specified region.
+func (c *awsClientsV2) getAWSConfigV2ForRegion(ctx context.Context, region string, opts awsOptions) (*aws.Config, error) {
+	if err := opts.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cacheKey := awsConfigCacheKey{
+		region:      region,
+		integration: opts.integration,
+	}
+
+	config, err := utils.FnCacheGet(ctx, c.configsCache, cacheKey, func(ctx context.Context) (*aws.Config, error) {
+		// TODO support integration.
+		if opts.credentialsSource == credentialsSourceIntegration {
+			return nil, trace.NotImplemented("missing aws integration config provider")
+		}
+
+		slog.DebugContext(ctx, "Initializing AWS config using ambient credentials.", "region", region)
+		config, err := awsAmbientConfigV2Provider(ctx, region, nil /*credProvider*/)
+		return config, trace.Wrap(err)
+	})
+	// TODO handle opts.customRetryer and opts.maxRetries
+	return config, trace.Wrap(err)
+}
+
+// getAWSConfigV2ForRole returns AWS config for the specified region and role.
+func (c *awsClientsV2) getAWSConfigV2ForRole(ctx context.Context, region string, options awsOptions) (*aws.Config, error) {
+	if err := options.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if options.baseConfigV2 == nil {
+		return nil, trace.BadParameter("missing base config")
+	}
+
+	cacheKey := awsConfigCacheKey{
+		region:      region,
+		integration: options.integration,
+		roleARN:     options.assumeRoleARN,
+		externalID:  options.assumeRoleExternalID,
+	}
+	return utils.FnCacheGet(ctx, c.configsCache, cacheKey, func(ctx context.Context) (*aws.Config, error) {
```

Comment:
there's a `CredentialCache` in aws-sdk-go-v2/aws, I wonder if we could replace our custom caching solution with that? I remember in the v1 api there was some issue with their caching, but I can't remember what it was that motivated us to use our own caching instead.